test/unix/test_lwt_bytes.ml

```diff
@@ -585,111 +606,63 @@ let suite = suite "lwt_bytes" [
       in
-      let () = Lwt_main.run (Lwt.join [client (); server ()]) in
+      let () = udp_server_client_exchange server_logic client_logic in
       let check = "abcdef" = Lwt_bytes.to_string buf in
       Lwt.return check
     end;
 
     test "bytes recv_msg" ~only_if:(fun () -> not Sys.win32) begin fun () ->
-      let server_is_ready, notify_server_is_ready = Lwt.wait () in
       let buffer = Lwt_bytes.create 6 in
       let offset = 0 in
       let io_vectors = [Lwt_bytes.io_vector ~buffer ~offset ~length:6] in
-      let server () =
-        let sock = Lwt_unix.socket Lwt_unix.PF_INET Lwt_unix.SOCK_DGRAM 0 in
-        let sockaddr = Lwt_unix.ADDR_INET (Unix.inet_addr_loopback, 0) in
-        Lwt_unix.bind sock sockaddr
-        >>= fun () ->
-        let server_address = Lwt_unix.getsockname sock in
-        Lwt.wakeup_later notify_server_is_ready server_address;
-        Lwt_bytes.recv_msg ~socket:sock ~io_vectors
-        >>= fun (_n, _sockaddr) -> Lwt_unix.close sock
+      let server_logic socket =
+        Lwt_bytes.recv_msg ~socket ~io_vectors
       in
-      let client () =
-        server_is_ready
-        >>= fun sockaddr ->
-        let sock = Lwt_unix.socket Lwt_unix.PF_INET Lwt_unix.SOCK_DGRAM 0 in
+      let client_logic socket sockaddr =
         let message = Lwt_bytes.of_string "abcdefghij" in
-        Lwt_bytes.sendto sock message 0 9 [] sockaddr
-        >>= fun (_n) -> Lwt_unix.close sock
+        Lwt_bytes.sendto socket message 0 9 [] sockaddr
       in
-      let () = Lwt_main.run (Lwt.join [client (); server ()]) in
+      let () = udp_server_client_exchange server_logic client_logic in
       let check = "abcdef" = Lwt_bytes.to_string buffer in
       Lwt.return check
     end;
 
     test "bytes send_msg" ~only_if:(fun () -> not Sys.win32) begin fun () ->
-      let server_is_ready, notify_server_is_ready = Lwt.wait () in
       let buffer = Lwt_bytes.create 6 in
       let offset = 0 in
-      let io_vectors = [Lwt_bytes.io_vector ~buffer ~offset ~length:6] in
-      let server () =
-        let sock = Lwt_unix.socket Lwt_unix.PF_INET Lwt_unix.SOCK_DGRAM 0 in
-        let sockaddr = Lwt_unix.ADDR_INET (Unix.inet_addr_loopback, 0) in
-        Lwt_unix.bind sock sockaddr
-        >>= fun () ->
-        let server_address = Lwt_unix.getsockname sock in
-        Lwt.wakeup_later notify_server_is_ready server_address;
-        Lwt_bytes.recv_msg ~socket:sock ~io_vectors
-        >>= fun (_n, _sockaddr) -> Lwt_unix.close sock
+      let server_logic socket =
+        let io_vectors = [Lwt_bytes.io_vector ~buffer ~offset ~length:6] in
+        Lwt_bytes.recv_msg ~socket ~io_vectors
       in
-      let client () =
-        server_is_ready
-        >>= fun sockaddr ->
-        let sock = Lwt_unix.socket Lwt_unix.PF_INET Lwt_unix.SOCK_DGRAM 0 in
-        Lwt_unix.connect sock sockaddr
+      let client_logic socket sockaddr =
+        Lwt_unix.connect socket sockaddr
```

Comment:
Is this needed for the UDP socket? I have no experience with `send_msg`, but this looks suspicious.

Reply:
Unlike with `Lwt_bytes.sendto socket message 0 9 [] sockaddr`, I need to "link" the socket to an address before calling `Lwt_bytes.send_msg ~socket ~io_vectors ~fds:[]`.

Furthermore according to the manpage of connect http://man7.org/linux/man-pages/man2/connect.2.html connect can be used with UDP socket too:

> If the socket sockfd is of type SOCK_DGRAM, then addr is the address
       to which datagrams are sent by default, and the only address from
       which datagrams are received.  If the socket is of type SOCK_STREAM
       or SOCK_SEQPACKET, this call attempts to make a connection to the
       socket that is bound to the address specified by addr.

I have to say that the `Lwt_bytes.send_msg` do not work like the  C system call : 

```
ssize_t sendmsg(int sockfd, const struct msghdr *msg, int flags);
```
so maybe there is a better solution.